### DISCUSSION TITLE
Image embedded directly in the layout

### DISCUSSION THREAD

**Discussion by MuriloIPM** [eclipse-birt/birt#1175] [Q&A] [answered]
How do I insert a base64 image embedded directly in the layout in the backgroundImage property?

I need my images to be inside the report xml, to use as background in some rows

**Comment by wimjongman** [comment]
Did you google it? https://www.google.com/search?q=eclipse+birt+images+bas64

**Reply by MuriloIPM** [reply]
Yes, I only found it referring to the image tag but not the backgroundImage property.

I need to set in: "<property name="backgroundImage">Referencing bas64 image here</property>"

**Reply by MuriloIPM** [reply]
I found other related errors at: 
- https://forums.opentext.com/forums/developer/discussion/53486/embedded-background-images
- https://jazz.net/forum/questions/160194/birt-432-embedded-background-image-is-not-appearing-in-pdf-report

**Reply by speckyspooky** [reply]
I have done some basic tests with embedded images with grid-rows and table-rows
and main output types HTML and PDF. My result is that I can the the background image on both outputs.

![BIRT_background_image rptdesign](https://user-images.githubusercontent.com/41593722/211200730-ca40e040-357d-423a-ab0c-b9b55d55547b.png)


I will add my test report and the PDF-document. We would need more details or a test report from your side with your scripts.

** pdf document **
[BIRT-background_image_grid_table_row.pdf](https://github.com/eclipse/birt/files/10368603/BIRT-background_image_grid_table_row.pdf)

** test report **
[background_image.zip](https://github.com/eclipse/birt/files/10368604/background_image.zip)

**Reply by MuriloIPM** [reply]
Your report looks like this in my Birt Report Designer:
![image](https://user-images.githubusercontent.com/113636465/211296260-d6416e13-12ba-44e4-9510-79d7fd40ab90.png)
![image](https://user-images.githubusercontent.com/113636465/211296576-c14f96e6-2497-4f5f-8fb7-848758b0ba2e.png)
I switched to Embedded image as per the screenshot and it didn't work
Here's my test:
[MyReport_BackgroundEmbeded..rptdesign.zip](https://github.com/eclipse/birt/files/10372695/MyReport_BackgroundEmbeded.rptdesign.zip)

**Comment by claesrosell** [comment]
Are you sure that you are not encountering issue #657? If so, you will need Birt 4.9 or higher

**Reply by MuriloIPM** [reply]
No, but I took a look at the issue and it's not what I need, I need to change the background image to an embedded image in the layout.
Sorry i'm new with birt

**Comment by speckyspooky** [comment, marked as answer]
I come back with summary of analysis in the last 2 days.

**Summary at the beginning:**
Currently it isn't possible to use "Embedded Images" or data-URL like background image.
This point is given on all elements which use background images.

The pont is that the functionality wasn't programmed out in all option. The background image handling includes currently only the URL-image-usage based on local path at all. Yes the option of "Embedded Image" is selectable but it doesn't work.
I investigated on the code an there is no handling implemented.

This task is open since the implementation of background images. I figured out an bugzilla-request at the year 2012
**#386245** [https://bugs.eclipse.org/bugs/show_bug.cgi?id=386245](url)

The ticket includes also a change from 2012 and I tested the implementation in a local branch
but the change is more a hard fix which works only fpr master-page background and not on table-row-level.
In addition to it there are some smaller changes necessary to get it runable.

The next point which I figured out is that not all emitter works with the change of #386245 and doesn't work too with data-URL.
The emitter of HTML, office formats and open document formats must be changed.  

And the last point is that the handling of the background image is completely different implemented like the image-object.
This is caused due to the background-image location on the according elements. The background image isn't an own element according to image-element. The background images are only "properties" at the according structure level.

For my testing I changed:
- ReportElementEditPart.java (designer visibility)
- PageArea.java (handle the master page)
- BackgroundImageInfo.java (background image: this class implements the image, property backgroundImageType won't be handled)

Hint of "Embedded Image": This means the embeding is on the report level and not on script lavel.

To get the background image working in full are several changes required through the full BIRT-framework:
1. new implementation of the "background image type" usage (currently all my test crashed successfully) 
2. enhance back ground image, create an handling of the variantly data usage of "Embedded Image"
3. enhance the container elements which can use background images
4. enhance the emitter to display the background correctly

Such a "small" requirement but so many stuff to get it in work.
Currently you can only use the URL-image reference of the background images.

**Reply by MuriloIPM** [reply]
Thank you very much, I understood, sorry if I was not clear at some point.

**Reply by wimjongman** [reply]
@speckyspooky are you interested in becoming a committer again? We would love to have you back on the team.

**Reply by wimjongman** [reply]
@speckyspooky (I assume you are a former team member ;)

**Reply by speckyspooky** [reply]
@MuriloIPM 
Thank for your question. It is important to see which requirements are necessary to improve the product.
I will take a look into it.

My focus of the enhancement will be:
- elements: Master page beckground
- output / formats: HTML (Preview), PDF & PPTX/ODP

The background wasn't supported till now on MS Word, so it has no priority.
(Excel is out of scope completely.)

**Reply by speckyspooky** [reply]
@wimjongman This week I signed the commiter agreement to join to your dev team :o)
And thanks for your last comment but I wasn't an origin developer but I work with BIRT many times.

I created an own branch for my changes because the enhancement go trough many classes.
If I have a first solution I will let you know.

**Reply by wimjongman** [reply]
@speckyspooky I don't see your committer nomination. Did you mean signing the ECA? https://www.eclipse.org/legal/ECA.php

**Reply by speckyspooky** [reply]
@wimjongman Could you please advise me with the nomination and with our discussion which has impacts on my next steps. Thanks in advance to you!

**Reply by wimjongman** [reply]
Please write a mail to wim.jongman at gmail. We can discuss there.

**Comment by speckyspooky** [comment]
@MuriloIPM Short update with BIRT4.13 the background images will be supported with embedded images and also the data-URL will be part of the support.

**Reply by MuriloIPM** [reply]
Thanks for help me
